Debug build CI job crashes on `stats._unuran` threading test

Observed in gh-14962 (unrelated PR). Details from the [CI log](https://github.com/scipy/scipy/runs/4078367919?check_suite_focus=true):

```
sx...................................................................... [ 99%]
......................[gw0] node down: Not properly terminated
F
replacing crashed worker gw0
........................................................ss...ssssssssss. [ 99%]
.s...........................sssssssssssssssssssssssssssssssssssssssssss [ 99%]
ssssssssssssssssys:1: ResourceWarning: unclosed file <_io.TextIOWrapper name=0 mode='r' encoding='UTF-8'>
sys:1: ResourceWarning: unclosed file <_io.TextIOWrapper name=0 mode='r' encoding='UTF-8'>

=================================== FAILURES ===================================
_ build/testenv/lib/python3.8/site-packages/scipy/stats/tests/test_sampling.py _
[gw0] linux -- Python 3.8.0 /usr/bin/python3.8-dbg
worker 'gw0' crashed while running 'build/testenv/lib/python3.8/site-packages/scipy/stats/tests/test_sampling.py::test_threading_behaviour'
```

Likely a bug (threading behavior is something that could indeed cause this segfault), but it's not happening on other PRs. Let's just see if it happens again.

This test came in in commit 402ac3535b8c (main PR that added UNU.RAN, Cc @tirthasheshpatel).

This happened once more ([log](https://github.com/scipy/scipy/runs/4675703557?check_suite_focus=true)), again on the `Python38-dbg` job. Output:

```
..sssssssssssssssssssssssssssssssssssssssss [ 99%]
ssssssssssssssssssssssssssssssssssssssssssssssssssssssssssys:1: ResourceWarning: unclosed file <_io.TextIOWrapper name=0 mode='r' encoding='UTF-8'>
sys:1: ResourceWarning: unclosed file <_io.TextIOWrapper name=0 mode='r' encoding='UTF-8'>

=================================== FAILURES ===================================
_ build/testenv/lib/python3.8/site-packages/scipy/stats/tests/test_sampling.py _
[gw0] linux -- Python 3.8.0 /usr/bin/python3.8-dbg
worker 'gw0' crashed while running 'build/testenv/lib/python3.8/site-packages/scipy/stats/tests/test_sampling.py::test_threading_behaviour'
...
=========================== short test summary info ============================
FAILED scipy/stats/tests/test_sampling.py::test_threading_behaviour
...
= 1 failed, 35655 passed, 2599 skipped, 114 xfailed, 7 xpassed in 662.35s (0:11:02) =
```
The crash seems to occur very infrequently.

--- a/scipy/stats/_unuran/unuran_wrapper.pyx.templ
+++ b/scipy/stats/_unuran/unuran_wrapper.pyx.templ
@@ -430,8 +430,9 @@ cdef class Method:
         if self.urng == NULL:
             raise UNURANError(self._messages.get())
         self._check_errorcode(unur_set_urng(self.par, self.urng))
+        has_callback_wrapper = (self._callback_wrapper is not None)
         try:
-            if self._callback_wrapper is not None:
+            if has_callback_wrapper:
                 init_unuran_callback(&callback, self._callback_wrapper)
             self.rng = unur_init(self.par)
             # set self.par = NULL because a call to `unur_init` destroys
@@ -445,7 +446,7 @@ cdef class Method:
             unur_distr_free(self.distr)
             self.distr = NULL
         finally:
-            if self._callback_wrapper is not None:
+            if has_callback_wrapper:
                 release_unuran_callback(&callback)
 
     @cython.boundscheck(False)
@@ -465,12 +466,14 @@ cdef class Method:
             size_t i
             size_t size = len(out)
 
+        has_callback_wrapper = (self._callback_wrapper is not None)
+
         _lock.acquire()
         try:
             self._messages.clear()
             unur_set_stream(self._messages.handle)
 
-            if self._callback_wrapper is not None:
+            if has_callback_wrapper:
                 init_unuran_callback(&callback, self._callback_wrapper)
             for i in range(size):
                 out[i] = unur_sample_cont(rng)
@@ -481,7 +484,7 @@ cdef class Method:
                 raise UNURANError(msg)
         finally:
             _lock.release()
-            if self._callback_wrapper is not None:
+            if has_callback_wrapper:
                 release_unuran_callback(&callback)
 
     @cython.boundscheck(False)
@@ -501,12 +504,14 @@ cdef class Method:
             size_t i
             size_t size = len(out)
 
+        has_callback_wrapper = (self._callback_wrapper is not None)
+
         _lock.acquire()
         try:
             self._messages.clear()
             unur_set_stream(self._messages.handle)
 
-            if self._callback_wrapper is not None:
+            if has_callback_wrapper:
                 init_unuran_callback(&callback, self._callback_wrapper)
             for i in range(size):
                 out[i] = unur_sample_discr(rng)
@@ -517,7 +522,7 @@ cdef class Method:
                 raise UNURANError(msg)
         finally:
             _lock.release()
-            if self._callback_wrapper is not None:
+            if has_callback_wrapper:
                 release_unuran_callback(&callback)
 
     def rvs(self, size=None, random_state=None):

